File: Base/Source/ExceptionHandler.vb
Imports System.Text

Public Class ExceptionHandler
    Shared Sub Handle(ex As Exception, Optional rethrow As Boolean = True)
        Handle(ex, TraceEventType.Information, "", rethrow)
    End Sub

    Shared Sub Handle(ex As Exception, severity As TraceEventType, additionalInfo As String, Optional rethrow As Boolean = True)
        If ex IsNot Nothing Then
            My.Application.Log.WriteEntry($"Exception type {ex.GetType().Name}", severity)
            Log(ex, severity, additionalInfo)
            Dim inner = ex.InnerException
            While inner IsNot Nothing
                My.Application.Log.WriteEntry($"Exception type {ex.InnerException.GetType().Name}", severity)
                Log(ex.InnerException, severity, additionalInfo)
                inner = inner.InnerException
            End While
            If rethrow Then
                Throw ex
            End If
        End If
    End Sub

    Private Shared Sub Log(ex As Exception, severity As TraceEventType, additionalInfo As String)
        My.Application.Log.WriteEntry("", severity)
        My.Application.Log.WriteException(ex, severity, additionalInfo)
        My.Application.Log.WriteEntry(ex.StackTrace, severity)
        My.Application.Log.TraceSource.Flush()
    End Sub

    Shared Function GetString(ex As Exception) As String
        Dim msg As New StringBuilder
        msg.AppendLine(ex.Message)
        msg.AppendLine(ex.StackTrace)
        If ex.InnerException IsNot Nothing Then
            msg.AppendLine(ex.InnerException.Message)
            msg.AppendLine(ex.InnerException.StackTrace)
        End If
        Return msg.ToString
    End Function

    Shared Function GetAllMessages(ex As Exception) As String
        Dim msg As New StringBuilder
        msg.AppendLine(ex.Message)
        Dim inner = ex.InnerException
        While inner IsNot Nothing
            msg.AppendLine(inner.Message)
            inner = inner.InnerException
        End While
        Return msg.ToString
    End Function
End Class

File: Base/src/ExceptionHandler.cs
using NLog;
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using Wells.Utilities.LoggerService;

namespace Wells.Base
{
    public static class ExceptionHandler
    {

        private static LoggerManager loggerManager;

        public static LoggerManager LoggerManager
        {
            get
            {
                if (loggerManager == null)
                {
                    LogManager.LoadConfiguration(Path.Combine(Directory.GetCurrentDirectory(), "nlog.config"));
                    loggerManager = new LoggerManager();
                }
                return loggerManager;
            }
        }

        public static void Handle(Exception ex)
        {
            Handle(ex, TraceEventType.Error, "", true);
        }

        public static void Handle(Exception ex, bool rethrow)
        {
            Handle(ex, TraceEventType.Error, "", rethrow);
        }

        public static void Handle(Exception ex, TraceEventType severity, string additionalInfo)
        {
            Handle(ex, TraceEventType.Error, "", true);
        }

        public static void Handle(Exception ex, TraceEventType severity, string additionalInfo, bool rethrow)
        {
            if (ex != null)
            {
                LoggerManager.Log($"Exception type {ex.GetType().Name}", severity);
                Log(ex, severity, additionalInfo);
                var inner = ex.InnerException;
                while (inner != null)
                {
                    LoggerManager.Log($"Exception type {ex.InnerException.GetType().Name}", severity);
                    Log(ex.InnerException, severity, additionalInfo);
                    inner = inner.InnerException;
                }
                if (rethrow) { throw ex; }
            }
        }



        private static void Log(Exception ex, TraceEventType severity, string additionalInfo)
        {
            LoggerManager.Log("", severity);
            LoggerManager.Log(additionalInfo, severity);
            LoggerManager.Log(ex.StackTrace, severity);
        }

        public static string GetString(Exception ex)
        {
            var msg = new StringBuilder();
            msg.AppendLine(ex.Message);
            msg.AppendLine(ex.StackTrace);
            if (ex.InnerException != null)
            {
                msg.AppendLine(ex.InnerException.Message);
                msg.AppendLine(ex.InnerException.StackTrace);
            }
            return msg.ToString();
        }


        public static string GetAllMessages(Exception ex)
        {
            var msg = new StringBuilder();
            msg.AppendLine(ex.Message);
            var inner = ex.InnerException;
            while (inner != null)
            {
                msg.AppendLine(inner.Message);
                inner = inner.InnerException;
            }

            return msg.ToString();
        }
    }
}
